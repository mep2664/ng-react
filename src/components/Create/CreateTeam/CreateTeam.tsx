import * as React from "react";
import { Loader, SelectInput, TextInput } from "../..";
import styled from "styled-components";
import gql from "graphql-tag";
import { useMutation, useQuery } from "@apollo/react-hooks";

const GET_OPTIONS = gql`
query {
    allProjects {
        projectId
        projectName
    }
    allUsers {
        userId
        firstName
        lastName
    }
}
`;

const CREATE_TEAM = gql`
    mutation CREATE_TEAM($teamName:String!, $members:[ID], $projects:[ID], $status:String, $dateCreated:DateTime) {
        createTeam(teamName:$teamName, members:$members, projects:$projects, status:$status, dateCreated:$dateCreated){
            team {
                teamId
                teamName
                status
                members
                projects
                dateCreated
            }
        }
    }
`;

const SelectInputs = styled.div`
    display: grid;
    grid-gap: 10px;
    align-items: center;
    justify-items: end;
    grid-template-columns: auto minmax(auto, 300px);
`;

const statusOptions = [
    { caption: "Pending", value: 1 },
    { caption: "Active", value: 2 },
    { caption: "Terminated", value: 3 },
];

export const CreateTeam: React.FC = () => {
    const [teamName, setTeamName] = React.useState<string>("");
    const [members, setMembers] = React.useState<string[]>([]);
    const [projects, setProjects] = React.useState<string[]>([]);
    const [status, setStatus] = React.useState<string>("");
    const { loading, error, data } = useQuery(GET_OPTIONS);
    const [createTeam] = useMutation(CREATE_TEAM);

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const dateCreated = new Date().toISOString()
        const members_obj = members.map((member) => { return { userId: member } });
        console.log(members_obj);
        console.log(projects);
        const data = {
            teamName,
            members: members_obj,
            projects,
            status,
            dateCreated,
        }
        console.log(data);
        createTeam({ variables: data });

    }

    if (loading) {
        return <Loader />;
    }

    if (error) {
        return <div>{error.message}</div>;
    }

    const projectOptions = data.allProjects.map((project: any) => { return { caption: project.projectName, value: project.projectId } });
    const userOptions = data.allUsers.map((user: any) => { return { caption: `${user.firstName} ${user.lastName}`, value: user.userId } });

    console.log(projects);

    return (
        <div>
            {`Create Team...`}
            <form action="localhost:5556/team" method="post" onSubmit={handleSubmit}>
                <TextInput label="Team Name" name="teamName" onChange={(e: React.ChangeEvent<HTMLInputElement>) => setTeamName(e.target.value)} />
                <SelectInputs>
                    { /* TODO - Multiple Select Input With Working State */}
                    <SelectInput isMultiple={true} label="Team Members" name="members" options={userOptions} value={members}
                        onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setMembers(members.includes(e.target.value) ? members.splice(members.indexOf(e.target.value), 1) : members.concat(e.target.value))} />
                    <SelectInput label="Projects" name="projects" options={projectOptions} value={projects[0]} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setProjects([e.target.value])} />
                    <SelectInput label="Status" name="status" options={statusOptions} value={status} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setStatus(e.target.value)} />
                </SelectInputs>
                <input type="submit" value="submit" />
            </form>
        </div>
    );
}