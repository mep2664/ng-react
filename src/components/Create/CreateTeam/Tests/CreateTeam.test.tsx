import React from "react"
import { render, fireEvent, wait } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect";
import { CreateTeam, CREATE_TEAM } from "../CreateTeam"
import { MockedProvider } from '@apollo/react-testing';

const today = new Date();
const dateCreated = new Date(`${today.getMonth()}-${today.getDate()}-${today.getFullYear()}`).toISOString();
const vars = { teamName: "TEAM_ID", status: "Active" };
let numTimesSubmitted = 0;
const mocks = [
    {
        request: {
            query: CREATE_TEAM,
            variables: { ...vars, dateCreated },
        },

        result: () => {
            numTimesSubmitted++;
            return {
                data: {
                    createTeam: {
                        team: {
                            teamId: "TEAM_ID",
                            teamName: vars.teamName,
                            status: vars.status,
                            dateCreated,
                        },
                    },
                },
            };
        },
    },
];

describe("Tests for the CreateTeam component", () => {
    test("Component matches snapshot", async () => {
        const { container } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTeam />
            </MockedProvider>,
        );
        await wait();
        expect(container).toMatchSnapshot();
    });

    test("Changes team name correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTeam />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Team Name");
        fireEvent.change(input, { target: { value: vars.teamName } });
        expect(input).toHaveValue(vars.teamName);
    });

    test("Changes status correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTeam />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Status");
        fireEvent.change(input, { target: { value: vars.status } });
        expect(input).toHaveValue(vars.status);
    });

    test("Submits correctly", async () => {
        const { getByText, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTeam />
            </MockedProvider>,
        );
        await wait();

        const teamName = getByLabelText("Team Name");
        fireEvent.change(teamName, { target: { value: vars.teamName } });

        const status = getByLabelText("Status");
        fireEvent.change(status, { target: { value: vars.status } });

        const submit = getByText("submit");
        fireEvent.click(submit);
        await wait();
        expect(numTimesSubmitted).toEqual(1);
    });
});
