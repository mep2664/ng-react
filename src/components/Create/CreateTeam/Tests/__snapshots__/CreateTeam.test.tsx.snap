// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the CreateTeam component Component matches snapshot 1`] = `
<div>
  <div>
    Create Team...
    <form
      action="localhost:5557/team"
      method="post"
    >
      <label
        class="sc-iAyFgw fQEosV"
        for="textInput__Team Name"
      >
        Team Name
      </label>
      <input
        class="sc-hSdWYo fjzKFm"
        id="textInput__Team Name"
        name="teamName"
        type="text"
        value=""
      />
      <div
        class="sc-fMiknA byHEZU"
      >
        <label
          class="sc-eNQAEJ hALvCQ"
          for="selectInput__Status"
        >
          Status
        </label>
        <select
          class="sc-hMqMXs hplJYk"
          id="selectInput__Status"
          name="status"
        >
          <option
            value=""
          >
            &lt; Not Selected &gt;
          </option>
          <option
            value="Pending"
          >
            Pending
          </option>
          <option
            value="Active"
          >
            Active
          </option>
          <option
            value="Terminated"
          >
            Terminated
          </option>
        </select>
      </div>
      <input
        type="submit"
        value="submit"
      />
    </form>
  </div>
</div>
`;
